### PR TITLE
Generate report

### PR DESCRIPTION
**Pull request by jegorami**
Added endpoint for getting answers data for report generation. 
This PR depends on PR #7

### CODE REVIEW

**Review by tomijarvi — APPROVED** [review]
For me, this looks good 👍 @JsseL ?

File: assignments/tests/factories.py [hunk 1]
```diff
@@ -82,3 +82,31 @@ class Meta:
         model = models.SchoolClass
 
     name = factory.Sequence(lambda n: 'class_%s' % n)
+
+
+class StudentFactory(factory.DjangoModelFactory):
+    class Meta:
+        model = models.Student
```
Comment (JsseL) [inline assignments/tests/factories.py]: I guess we didn't change this to `Answer` [after all](https://github.com/City-of-Turku/living-environment/pull/7#discussion_r112890102).. Well, I guess `Student` is fine also.

File: assignments/serializers.py [hunk 2]
```diff
@@ -114,3 +116,70 @@ def save(self):
...
+            [(key.replace('student__', ''), value) for key, value in self.context['query_params'].items()])
+        students_per_school = Student.objects.filter(
+            school__assignment__slug=self.context['assignment_slug'],
+            **student_query).values('school__name').annotate(count=Count('id')).order_by()
+        students_per_class = Student.objects.filter(
+            school__assignment__slug=self.context['assignment_slug'],
+            **student_query).values('school_class__name').annotate(count=Count('id')).order_by()
+        return {
+            'per_school': students_per_school,
+            'per_class': students_per_class
```
Comment (JsseL) [inline assignments/serializers.py]: Note how odd this sounds `students_per_class` compared to `answers_per_class`. Clearly it's not the number of students in that given class, but it's a number of answers.

File: assignments/serializers.py [hunk 3]
```diff
@@ -114,3 +116,70 @@ def save(self):
...
+    class Meta:
+        model = Assignment
+        fields = ['name', 'sections', 'students']
+
+    def get_students(self, obj):
+        student_query = dict(
+            [(key.replace('student__', ''), value) for key, value in self.context['query_params'].items()])
+        students_per_school = Student.objects.filter(
+            school__assignment__slug=self.context['assignment_slug'],
+            **student_query).values('school__name').annotate(count=Count('id')).order_by()
```
Comment (JsseL) [inline assignments/serializers.py]: So the `**student_query` is used to filter answers for specific school and class?

This whole method feels a bit complicated. Could you move some of the logic into `Assignment` and try to have something like:
```
answers_per_school = Assignment.get_answers(school=school).count()
answers_per_school = Assignment.get_answers(school_class=school_class).count()
```
Then in `Assignment` model's `get_answers` you could encapsulate the more complex querying. Having such methods in `Assignment` makes the code also more reusable than having complex logic in specific serializer.

**Review by JsseL — COMMENTED** [review]
See the few suggestions about making the code bit more readable.

File: assignments/serializers.py (hunk 2 above)
Comment (jegorami) [reply assignments/serializers.py]: Couldn't find better name. This is something like single game play that all answers are related to

Comment (JsseL) [reply assignments/serializers.py]: Ah sorry, misread the code. It's a queryset of Students and not a number. So should it be `answers_by_school` and `answers_by_class`?

Comment (jegorami) [reply assignments/serializers.py]: It's how many students have played game sorted by school, and how many students have played the game sorted by class (school is not taken into consideration in that case).
But, that can be narrowed by teachers query(teacher can select school, class or both)

Comment (JsseL) [reply assignments/serializers.py]: So `answers_by_school` and  `ansers_by_class`, right?

**Comment by jegorami** [comment]
@JsseL I renamed Student to Session and correct code according to comments

File: assignments/models.py [hunk 4]
```diff
@@ -34,6 +34,15 @@ def __str__(self):
     def get_absolute_url(self):
         return reverse('assignments:assignment-detail', args=[self.slug])
 
+    def get_submissions(self, school=None, school_class=None):
+        """
+        Get all submissions related to the assignment filtered by school or school_class if given.
+        """
+        submissions = Submission.objects.all()
+        submissions = submissions.filter(school=school) if school is not None else submissions
+        submissions = submissions.filter(school_class=school_class) if school_class is not None else submissions
```
Comment (JsseL) [inline assignments/models.py]: Don't like the inline pattern, but no need to fix it. I would just use:
```
if school_class:
    submissions = submissions.filter(school_class=school_class)
```

File: assignments/models.py [hunk 5]
```diff
@@ -75,6 +84,15 @@ class Meta:
         verbose_name = _('open text task')
         verbose_name_plural = _('open text tasks')
 
+    def get_answers(self, school=None, school_class=None):
+        """
+        Get open text answers filtered by school and class
+        """
+        answers = self.open_text_answers.all()
+        answers = answers.filter(submission__school__name=school) if school is not None else answers
+        answers = answers.filter(submission__school_class__name=school_class) if school_class is not None else answers
```
Comment (JsseL) [inline assignments/models.py]: For example here the line gets already really long.

**Comment by JsseL** [comment]
Starting to look good. :+1:

File: assignments/models.py (hunk 5 above)
Comment (jegorami) [reply assignments/models.py]: @JsseL I see what you mean. I removed inline pattern.

**Comment by jegorami** [comment]
Hope everything is fixed now, @JsseL

**Comment by JsseL** [comment]
Yep. LGTM. :+1:

**Pull request merged**